beps(dynamic-discovery): pivot to just plugin listing with static config (#25431)
backstage/backstage: beps/0005-definitive-plugin-listing/README.md

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this BEP. Describe why the change is important and the benefits to users.
-->

Split backends are a consistently difficult space to operate in and design for. There has been a growing desire for the framework to provide a way to get a list of the installed plugins. This was nearly impossible in the old backend, where plugins were hosted on denormalized routes and had non-standard startup sequences. In the new backend, this has become significantly more doable. Moving this forward would unblock a number of cases that require knowledge of your entire Backstage installation, namely a single OpenAPI spec for your instance, checking installed permissions, and DevTools information.

### Goals

<!--
List the specific goals of the BEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

> 1. As an integrator, I can now get a list of currently installed features across my deployment.
> 1. As an integrator/administrator, I can get a list of currently installed features for a single instance in my deployment.

Reviewer: ```suggestion
1. As a plugin builder, I can now get a list of currently installed features across my deployment.
1. As a plugin builder, I can get a list of currently installed features for a single instance in my deployment.
```

Right? Really the role that would be doing that. If you want to expose the available features for an integrator/operator you'd need to do that through a plugin.

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs or even code snippets.
If there's any ambiguity about HOW your proposal will be implemented, this is the place to discuss them.
-->

### Glossary

### New Type `BackendFeatureMeta`

```ts
type BackendFeatureMeta =
  | {
      type: 'plugin';
      pluginId: string;
    }
  | {
      type: 'module';
      pluginId: string;
      moduleId: string;
    };
```

### New `InstanceMetadataService`

While we could attach the existing information to the `PluginMetadataService`, we propose a new service that handles instance-level information. The existing `PluginMetadataService` should reveal information about the plugin itself, its `pluginId`, dependencies or similar. The new `InstanceMetadataService` should give you information about the entire Backstage instance that you're interrogating. One could imagine this service also having information about instance URLs, health or gateway status.

```ts
interface InstanceMetadataService {

>   getInstalledFeatures: () => BackendFeatureMeta[];

Reviewer: Related to the discussion above, the service that returns locally installed features would probably have a method that's called something like `getLocallyInstalledFeatures`, `getColocatedFeatures`, or `getFeaturesInstalledOnTheSameInstance`, something very explicit x)

Author: updated in [05d1f2b](https://github.com/backstage/backstage/pull/25431/commits/05d1f2b5902124dcf7a087b4e3abf8e29fb8eb27)

}
```

Currently, this would need to be bootstrapped directly in `BackendInitializer` or `BackstageBackend`. It may make sense to create a new `FeatureRegistry` class that can expose this instead similar to the `PluginMetadataService`.

### New HTTP API to expose `InstanceMetadataService`

We also propose exposing the new services through an HTTP API, so that these can be queried.

```yaml
paths:

>   /.backstage/instanceInfo/features/installed:

Reviewer: So to be clear, just features in this instance right?

Author: yes

    get:
      summary: Get a list of installed features for this instance.
      operationId: GetInstalledFeaturesByInstance
      responses:
        '200':
          description: Successful operation
          content:
            application/json:
              schema:
                type: array
                items:
                  oneOf:
                    - type: object
                      properties:
                        type:
                          type: string
                          enum:
                            - plugin
                        pluginId: { type: string }
                    - type: object
                      properties:
                        type:
                          type: string
                          enum:
                            - module
                        pluginId: { type: string }
                        moduleId: { type: string }
```

### New `DeploymentMetadataPlugin`

> This will be a new core plugin that will be provided through a new package. The idea is that it will show all information across your entire deployment. One can imagine this hosting aggregate information around instances in the deployment, like their internal and external URLs.

Reviewer: What do you think about skipping the plugin for now and just making it a core service that consumes config? Then we could add a plugin if needed in the future

Author: makes sense, done in [05d1f2b](https://github.com/backstage/backstage/pull/25431/commits/05d1f2b5902124dcf7a087b4e3abf8e29fb8eb27)

```ts
interface BackstageInstance {
  url: {
    internal: string;
    external: string;
  };
}

> interface DeploymentMetadataPluginExtensionPoint {

Reviewer: I think we'll want to provide these either with a custom implementation, or via configuration tbh, rather than mutating the service.

Author: updated in [05d1f2b](https://github.com/backstage/backstage/pull/25431/commits/05d1f2b5902124dcf7a087b4e3abf8e29fb8eb27)

1. The list of plugins will not be used in routing or discovery.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation.
-->

Most of the work below is API definitions for hosting the list of features and better abstractions to make that list available through HTTP. This is not intended to perform much heavy lifting, that can happen in a further design. We will build a default implementation on the existing static `HostDiscovery` config definition for split deployments. Users can override the API as they need.

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs or even code snippets.
If there's any ambiguity about HOW your proposal will be implemented, this is the place to discuss them.
-->

### Glossary

### New Type `BackstageFeature`

```ts
type BackstageFeature =
  | {
      type: 'plugin';
      pluginId: string;
    }
  | {

>       type: 'module';

Author: Definitely could include the extension points + service factories, I wasn't familiar with those so didn't include them here, but they likely fit in well.

Loaders could make sense to include in a future iteration, but they may make more sense as a separate API. Similarly, dependencies and their versions _could_ fit into this API or may make sense as a separate API that only includes versioned registrations - I'd imagine something like service factories are likely to be less well versioned.

>       type: 'module';

Reviewer: Let's keep it slim like this for now :>

      pluginId: string;
      moduleId: string;
    };
```

> ### New `InstanceMetadataService`

Reviewer 2: One more thought. Making it a service implies the ability to depend on it and possibly even replace it. Does either of those two actions make any sense? Would it be an antipattern to try to enumerate features inside another plugin, while early startup is still in full swing? Essentially - what if this was dialed back to only being the `instanceInfo` HTTP endpoint and no service at all? Would that make sense?

Reviewer: I think it makes sense to split into two services tbh, one that is strictly for colocated plugins and very explicit about that. That one would be provided by the backend initializer, just like the existing plugin metadata service. In addition to that we also add a separate service that is the one with more appealing naming and the one actually intended to be used for discovery. We keep that one replacable though, so that one can plugin in their own way of listing available services across all deployed instances.

Author: @Rugvip Splitting the `InstanceMetadata` service into 2 or using the services described here to create a `BackstageMetadataService` or something that could be used for discovery?

### New `DeploymentMetadataService`

This will be a new service that can be provided either as a core service or a new package. The idea is that it will show all information across your entire deployment. One can imagine this hosting aggregate information around instances in the deployment, like their internal and external URLs.

```ts
interface DeploymentMetadataService {
  getInstalledFeatures: () => BackstageFeature[];
}
```

By default, this will either
a. be the exact same as `InstanceMetadataService` if we are not using `HostDiscovery` static config, or
b. this will be pulled from that `HostDiscovery` static config.

That would look something like,

```yaml
discovery:
  endpoints:
    - target: https://internal.example.com/internal-catalog
      plugins: [catalog]
    - target: https://internal.example.com/secure/api/{{pluginId}}
      plugins: [auth, permission]
    - target:
        internal: https://internal.example.com/search
        external: https://example.com/search
      plugins: [search]
```

So, the above interface would return for the above config,

```ts
const installedFeatures = deploymentMetadataService.getInstalledFeatures();

> // installedFeatures = [{type: 'plugin', pluginId: 'catalog'}, {type: 'plugin', pluginId: 'auth'}, {type: 'plugin', pluginId: 'permission'}, {type: 'plugin', pluginId: 'search'}]

Author: As is, this implementation would only return plugins and nothing dynamic since it isn't calling into the individual instance. I'll call that out better here, I was trying to provide a solid static implementation that was extendable once we get more dynamic.

I was trying to scope this to a static implementation for ease of use and less config/routing problems. Happy to go down the route of URL based discovery, but I was thinking that would be a v2 of this BEP and not immediately implemented.

> // installedFeatures = [{type: 'plugin', pluginId: 'catalog'}, {type: 'plugin', pluginId: 'auth'}, {type: 'plugin', pluginId: 'permission'}, {type: 'plugin', pluginId: 'search'}]

Author: Coming back to this, this is a poor example that matches HostDiscovery but doesn't demonstrate really what I was trying to do here. `moduleId` is difficult to do without the dynamic discovery this BEP had originally. We could update this to URL based, ie initially using the list of URLs configured in the discovery config and working backwards based on their instance metadata endpoints - I would like that to be a v2 of this proposal, it definitely makes sense, but has more moving parts that I think implementing the static version of would help flush out.

> // installedFeatures = [{type: 'plugin', pluginId: 'catalog'}, {type: 'plugin', pluginId: 'auth'}, {type: 'plugin', pluginId: 'permission'}, {type: 'plugin', pluginId: 'search'}]

Author: Coming back to this again - I think URLs is the right answer here. It solves the issue with modules not showing up in my initial design and ensures that data will be consistent if configs are different across machines. I'll update the doc with that.

```

### New HTTP API to expose `InstanceMetadataService`

We also propose exposing the new services through an HTTP API, so that these can be queried.

```yaml
paths:
  /.backstage/instanceInfo/features/installed:
    get:
      summary: Get a list of installed features for this instance.
      operationId: GetInstalledFeaturesByInstance
      responses:
        '200':
          description: Successful operation
          content:
            application/json:
              schema:
                type: array
                items:
                  oneOf:
                    - type: object
                      properties:
                        type:
                          type: string
                          enum:
                            - plugin
                        pluginId: { type: string }
                    - type: object
                      properties:
                        type:
                          type: string
                          enum:

>                             - plugin

Reviewer 2: ```suggestion
                            - module
```

                        pluginId: { type: string }
                        moduleId: { type: string }
```

### New HTTP API to expose `DeploymentMetadataService`

```yaml
paths:

>   /.backstage/deploymentInfo/features/installed:

Reviewer 2: So this is more for frontend consumption, while the other one is for service consumption - so to speak? As in, if we make a frontend plugin that wants to show an overview of your Backstage ecosystem, would it just pick some random plugin ID of sorts and call this enpoint there? And then that would perhaps use discovery and make a bunch of instance info calls to do its job?

Makes me wonder if the deployment info service shouldn't be a service, but rather a core plugin whose ID you could resolve with discovery and make calls to.

Author: Interesting point, I definitely agree - instance data needs to be a service to tie into startup information, but this could be spun off to a separate plugin for sure.

    get:
      summary: Get a list of installed features for this deployment.
      operationId: GetInstalledFeaturesByDeployment
      responses:
        '200':
          description: Successful operation
          content:
            application/json:
              schema:
                type: array
                items:
                  oneOf:
                    - type: object
                      properties:
                        type:
                          type: string
                          enum:
                            - plugin
                        pluginId: { type: string }
                    - type: object
                      properties:
                        type:
                          type: string
                          enum:

>                             - plugin

Reviewer 2: ```suggestion
                            - module
```

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this BEP. Describe why the change is important and the benefits to users.
-->

Split backends are a consistently difficult space to operate in and design for. There has been a growing desire for the framework to provide a way to get a list of the installed plugins. This was nearly impossible in the old backend, where plugins were hosted on denormalized routes and had non-standard startup sequences. In the new backend, this has become significantly more doable. Moving this forward would unblock a number of cases that require knowledge of your entire Backstage installation, namely a single OpenAPI spec for your instance, checking installed permissions, and DevTools information.

> ### Goals

Reviewer: Feeling that this area is perhaps lacking a little bit, there's probably something around building towards enabling dynamic discovery too?

<!--
List the specific goals of the BEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

1. As a plugin builder, I can now get a list of currently installed features across my deployment.

> 1. As a plugin builder, I can get a list of currently installed features for a single instance in my deployment.

Reviewer: I realized this is probably not something that a plugin builder should need to care about? Maybe it's interesting for an admin though?

1. As an integrator/administrator, I can continue to use my static `HostDiscovery` config.

### Non-Goals

<!--
What is out of scope for this BEP? Listing non-goals helps to focus discussion
and make progress.
-->

> 1. The list of plugins will not change unless there is a config change, they will not be dynamically loaded.

Reviewer: From who's point of view is this? Either way it seems like features could be added or removed without a config change. Either in the local instance with a restart, or in a remote instance.

          description: Successful operation
          content:
            application/json:
              schema:
                type: array
                items:
                  oneOf:
                    - type: object
                      properties:
                        type:
                          type: string
                          enum:
                            - plugin
                        pluginId: { type: string }
                    - type: object
                      properties:
                        type:
                          type: string
                          enum:
                            - module
                        pluginId: { type: string }
                        moduleId: { type: string }
```

### New `SystemMetadataService`

This will be a new core service. The idea is that it will show information about your whole Backstage deployment, which can have multiple instances. Initially, this will host aggregate information around instances in the deployment, like their internal and external URLs.

```ts
interface BackstageInstance {
  url: {
    internal: string;
    external: string;
  };
}

interface SystemMetadataService {

>   getInstances: () => BackstageInstance[];

Author: ```suggestion
  getInstances: () => Promise<BackstageInstance[]>;
```

  };
}

> interface DeploymentMetadataService {

Reviewer: perhaps SystemMetadataService or such? I'm worried deployment is a bit overloaded in kubernetes where it really just means a single instance in these terms

>   listInstances: () => BackstageInstance[];

Reviewer: hmm, wonder if this should just be the same list of features rather than a list of instances. A slight issue I see with this is that we don't have the concept of routing to an instance, only plugins. It'd be an additional thing to add to the system.

Author: Yeah, I call that out as a TODO below in the doc. I'll make that a bit more clear in the BEP - here's the original thread with Freben around plugins vs instances, https://github.com/backstage/backstage/pull/25431#discussion_r1722788174.

IMO we should stick with instances here and rely on the instance metadata service HTTP API to retrieve all metadata, the plugin-first approach is more discovery specific and in my mind is the custom discovery implementation using these 2 APIs. The big issue with just scraping the existing static config + list of the current instance's plugins is that it requires the config to be up to date with everything that's deployed. Couple of examples where I think that is a little rough,
1. Let's say you deploy the catalog + scaffolder to their own pods, but you have a single instance running smaller plugins, ie api-docs, techdocs, user-settings, etc. That single instance will have the correct list of plugins for the entire system, but unless you add each pluginId to your configs for catalog + scaffolder, they'd be inaccurate.
2. Similarly, if you want to dynamically add a plugin to a deployment, assuming you're just using the default discovery implementation, you'd need to adjust the config as well.

I think having the system metadata service provide instances and let the clients scrape instances themselves provides a bridge between the purely static and a more dynamic system that the future may hold.

}
```

By default, this will
a. be the exact same as `InstanceMetadataService` if we are not using `HostDiscovery` static config, or

> b. use the `HostDiscovery` static config to fetch other instance's base URLs.

Reviewer: another option is to simply rely on local + explicitly configured plugins in discovery for this for now
